v2/yarpcgrpc/outbound.go

```diff
@@ -0,0 +1,318 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"context"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"go.uber.org/yarpc/v2"
+	intyarpcerror "go.uber.org/yarpc/v2/internal/internalyarpcerror"
+	"go.uber.org/yarpc/v2/yarpcerror"
+	"go.uber.org/yarpc/v2/yarpcpeer"
+	"go.uber.org/yarpc/v2/yarpctracing"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// UserAgent is the User-Agent that will be set for requests.
+// http://www.grpc.io/docs/guides/wire.html#user-agents
+const UserAgent = "yarpc-go/" + yarpc.Version
+
+var _ yarpc.UnaryOutbound = (*Outbound)(nil)
+
+// Outbound sends YARPC requests over gRPC. It is recommended that services use
+// a single HTTP dialer to construct all HTTP outbounds, ensuring efficient
+// sharing of resources across the different outbounds.
+type Outbound struct {
+	// Chooser is a peer chooser for outbound requests.
+	Chooser yarpc.Chooser
+
+	// Dialer is an alternative to specifying a Chooser. The outbound will dial
+	// the address specified in the URL.
+	Dialer yarpc.Dialer
+
+	// URL specifies the template for the URL this outbound makes requests to.
+	//
+	// For yarpc.Chooser-based outbounds, the peer (host:port) section of the URL
+	// may vary from call to call.
+	URL *url.URL
+
+	// Tracer attaches a tracer for the outbound.
+	Tracer opentracing.Tracer
+}
+
+// Call implements yarpc.UnaryOutbound#Call.
+func (o *Outbound) Call(ctx context.Context, req *yarpc.Request, reqBuf *yarpc.Buffer) (*yarpc.Response, *yarpc.Buffer, error) {
+	if req == nil {
+		return nil, nil, yarpcerror.InvalidArgumentErrorf("request for grpc outbound was nil")
+	}
+
+	start := time.Now()
+
+	var responseBody []byte
+	var responseMD metadata.MD
+	invokeErr := o.invoke(ctx, req, reqBuf, &responseBody, &responseMD, start)
+
+	responseHeaders, err := getApplicationHeaders(responseMD)
+	if err != nil {
+		return nil, nil, err
+	}
+	return &yarpc.Response{
+			Headers:          responseHeaders,
+			ApplicationError: metadataToIsApplicationError(responseMD),
+		},
+		yarpc.NewBufferBytes(responseBody),
+		invokeErr
+}
+
+func (o *Outbound) invoke(
+	ctx context.Context,
+	req *yarpc.Request,
+	reqBuf *yarpc.Buffer,
+	responseBody *[]byte,
+	responseMD *metadata.MD,
```

Comment:
it looks like the underlying type for `metadata.MD` is a map, so if the pointer logic is for the sake of mutating the data, this doesn't need to be a pointer.

Reply:
older code